src/simple_cache_client.rs

```diff
@@ -1873,6 +1873,133 @@ impl SimpleCacheClient {
+    ///     Order::Ascending,
+    ///     0..10,
+    ///     true,
+    /// ).await? {
+    ///     Some(elements) => match elements {
+    ///         Elements::ValuesWithScores(elements) => {
+    ///             for element in elements.elements {
+    ///                 println!("element: {:?} has score: {}", element.value, element.score);
+    ///             }
+    ///         }
+    ///         Elements::Values(_) => {
+    ///             panic!("should have returned elements with scores");
+    ///         }
+    ///     }
+    ///     None => {
+    ///         println!("sorted set or elements not found");
+    ///     }
+    /// }
+    /// # Ok(())
+    /// # })
+    /// # }
+    /// ```
+    pub async fn sorted_set_fetch_by_index(
+        &mut self,
+        cache_name: &str,
+        set_name: impl IntoBytes,
+        order: sorted_set::Order,
+        range: impl RangeBounds<i32>,
+        with_scores: bool,
+    ) -> MomentoResult<Option<crate::sorted_set::Elements>> {
+        // transforms the Rust `Range` start into a Momento start index. Because
+        // the Momento start index is always Inclusive (or Unbounded) we need to
+        // map the value of a Rust `Excluded` start bound to an
+        // `InclusiveStartIndex` by adding one to the value.
+        let start = match range.start_bound() {
+            std::ops::Bound::Included(v) => {
+                sorted_set_fetch_request::by_index::Start::InclusiveStartIndex(*v)
+            }
+            std::ops::Bound::Excluded(v) => {
+                sorted_set_fetch_request::by_index::Start::InclusiveStartIndex(*v + 1)
+            }
+            std::ops::Bound::Unbounded => {
+                sorted_set_fetch_request::by_index::Start::UnboundedStart(Unbounded {})
+            }
+        };
+
+        // transforms the Rust `Range` end into a Momento end index. Because the
+        // Momento end index is always Exclusive (or Unbounded) we need to map
+        // the value of a Rust `Included` end bound to an `ExclusiveEndIndex` by
+        // adding one to the value.
+        let end = match range.end_bound() {
+            std::ops::Bound::Included(v) => {
+                sorted_set_fetch_request::by_index::End::ExclusiveEndIndex(*v + 1)
+            }
+            std::ops::Bound::Excluded(v) => {
+                sorted_set_fetch_request::by_index::End::ExclusiveEndIndex(*v)
+            }
+            std::ops::Bound::Unbounded => {
+                sorted_set_fetch_request::by_index::End::UnboundedEnd(Unbounded {})
+            }
+        };
+
+        let request = self.prep_request(
+            cache_name,
+            SortedSetFetchRequest {
+                set_name: set_name.into_bytes(),
+                order: order.into(),
+                with_scores,
+                range: Some(sorted_set_fetch_request::Range::ByIndex(
+                    sorted_set_fetch_request::ByIndex {
+                        start: Some(start),
```

Comment:
Is there any downside to always providing Some start and Some end? I'm confused why we have an Option and then either some bound or the dedicated Unbounded type.

Reply:
not sure the cleanest way to map it to rust, but in the other SDKs, if you omit "start" it means "everything from the first index to the specified 'end'". If you omit "end" it means "everything from the specified 'start' to the last item in the set." That latter one in particular is important because the caller can't know for sure how many items are in the set when they make the call.

let me know if that answers your question?

Reply:
Yeah - so the question is more about the meaning in momento-protos. In Rust, the range type allows both start and end to be inclusive/exclusive/unbounded. It appears the momento sorted_set_fetch_request::ByIndex allows start to be Some(Inclusive) Some(Unbounded) or None. It's this duplication of the Some(Unbounded) and None that is a bit confusing in the underlying protos library. We're hiding it from the users of this crate, but it's unclear if Some(Unbounded) or None is a preferred representation when building the request.

Reply:
got it, thanks for the clarification. In the other SDKs we are sending `Unbounded`:

https://github.com/momentohq/client-sdk-javascript/blob/main/packages/client-sdk-nodejs/src/internal/data-client.ts#L2133

Reply:
but if it behaves the way you expect with explicit `Some`, I'm not aware of a downside. At the end of the day all we care about is that we can write integration tests for all of the expected behaviors and they pass :)

Reply:
Updated the example / doctest and it works fine.